File: python/test/t_Cylinder_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otmeshing as otm
import os

ot.TESTPREAMBLE()

# Test default constructor
print("--- Default constructor ---")
cyl_default = otm.Cylinder()
assert cyl_default.getDimension() == 0
print("OK")

# Test parameter constructor with a simple square base (exact geometry)
print("--- Cylinder with square base ---")
base_mesh = ot.IntervalMesher([5, 5]).build(ot.Interval([-1.0, -1.0], [1.0, 1.0]))
extension = ot.Interval([-2.0], [2.0])
injection = [2]
M = 3
cyl_square = otm.Cylinder(base_mesh, extension, injection, M)

assert cyl_square.getDimension() == 3
ott.assert_almost_equal(
    cyl_square.getVolume(), base_mesh.getVolume() * extension.getVolume()
)

# check base
retrieved_base = cyl_square.getBase()
ott.assert_almost_equal(retrieved_base.getVertices(), base_mesh.getVertices())
assert retrieved_base.getSimplices() == base_mesh.getSimplices()

# check extension
retrieved_ext = cyl_square.getExtension()
assert retrieved_ext.getLowerBound()[0] == -2.0
assert retrieved_ext.getUpperBound()[0] == 2.0

# check injection
retrieved_inj = cyl_square.getInjection()
assert retrieved_inj == injection

# check discretization
assert cyl_square.getDiscretization() == M
print("OK")

# Test getVertices
print("--- getVertices ---")
ext_vertices = ot.IntervalMesher([M]).build(extension).getVertices()
vertices = cyl_square.getVertices()
assert vertices.getSize() == base_mesh.getVerticesNumber() * ext_vertices.getSize()
assert vertices.getDimension() == 3
print("OK")

# Test getBoundingBox
print("--- getBoundingBox ---")
bbox = cyl_square.getBoundingBox()
ott.assert_almost_equal(bbox.getLowerBound(), [-1.0, -1.0, -2.0])
ott.assert_almost_equal(bbox.getUpperBound(), [1.0, 1.0, 2.0])
print("OK")

# Test isConvex
print("--- isConvex ---")
assert cyl_square.isConvex()
print("OK")

# Test computeMesh
print("--- computeMesh ---")
computed_mesh = cyl_square.computeMesh()
assert computed_mesh.getVerticesNumber() > 0
print("OK")

# Test string representation
print("--- __repr__ / __str__ ---")
assert cyl_square.__repr__() == "class=Cylinder"
assert cyl_square.__str__() == "class=Cylinder"
print("OK")


# Test save/load
print("--- save/load ---")

study_file = "cylinder_test.xml"
study = ot.Study(study_file)
study.add("cylinder", cyl_square)
study.save()

loaded = otm.Cylinder()
study2 = ot.Study(study_file)
study2.load()
study2.fillObject("cylinder", loaded)
assert loaded.getDimension() == cyl_square.getDimension()
ott.assert_almost_equal(loaded.getVolume(), cyl_square.getVolume())
ott.assert_almost_equal(
    loaded.getBoundingBox().getLowerBound(), cyl_square.getBoundingBox().getLowerBound()
)
ott.assert_almost_equal(
    loaded.getBoundingBox().getUpperBound(), cyl_square.getBoundingBox().getUpperBound()
)
assert loaded.getDiscretization() == cyl_square.getDiscretization()
os.remove(study_file)
print("OK")

# Test non-convex cylinder
print("--- Non-convex cylinder ---")
snake_polyline = [
    [0.0, 0.0],
    [0.0, 5.0],
    [6.0, 5.0],
    [6.0, 0.0],
    [2.0, 0.0],
    [2.0, 3.0],
    [4.0, 3.0],
    [4.0, 2.0],
    [3.0, 2.0],
    [3.0, 1.0],
    [5.0, 1.0],
    [5.0, 4.0],
    [1.0, 4.0],
    [1.0, 0.0],
]
snake_mesh = otm.PolygonMesher().build(snake_polyline)
nonconvex_cyl = otm.Cylinder(snake_mesh, extension, injection, M)
assert nonconvex_cyl.getDimension() == 3
assert not nonconvex_cyl.isConvex()
ott.assert_almost_equal(
    nonconvex_cyl.getVolume(), snake_mesh.getVolume() * extension.getVolume()
)
print("OK")

# Test 1D base (extension in 2D)
print("--- 1D base cylinder ---")
base_1d = ot.IntervalMesher([5]).build(ot.Interval([-1.0], [1.0]))
ext_2d = ot.Interval([-1.0] * 2, [1.0] * 2)
inj_2d = [1, 2]
M2 = 2
cyl_1d = otm.Cylinder(base_1d, ext_2d, inj_2d, M2)
assert cyl_1d.getDimension() == 3
ott.assert_almost_equal(cyl_1d.getVolume(), base_1d.getVolume() * ext_2d.getVolume())
assert cyl_1d.isConvex()
bbox_1d = cyl_1d.getBoundingBox()
ott.assert_almost_equal(bbox_1d.getLowerBound(), [-1.0, -1.0, -1.0])
ott.assert_almost_equal(bbox_1d.getUpperBound(), [1.0, 1.0, 1.0])
print("OK")

# Test error cases
print("--- Error cases ---")
# Invalid injection size
with ott.assert_raises(TypeError):
    otm.Cylinder(base_mesh, extension, [2, 3], M)
print("Caught invalid injection size OK")

# Injection index out of range
with ott.assert_raises(TypeError):
    otm.Cylinder(base_mesh, extension, [5], M)
print("Caught invalid injection index OK")

# Keep existing disc-based tests
print("--- Disc-based cylinder ---")
N = 20
dim = 3
xc1 = 0.0
yc1 = 0.0
R1 = 1.0
H1 = 4.0
f1 = ot.SymbolicFunction(["x", "y"], [f"(x-({xc1}))^2+(y-({yc1}))^2"])
levelSet1 = ot.LevelSet(f1, ot.LessOrEqual(), R1**2)
base1 = ot.LevelSetMesher([N] * 2).build(
    levelSet1, ot.Interval([xc1 - R1, yc1 - R1], [xc1 + R1, yc1 + R1])
)
extension1 = ot.Interval([-H1 / 2] * (dim - 2), [H1 / 2] * (dim - 2))
injection1 = list(range(2, dim))
cyl1 = otm.Cylinder(base1, extension1, injection1, M)
assert cyl1.getDimension() == dim
assert cyl1.isConvex()
# volume = base_area * 4 (extension length)
vol1 = cyl1.getVolume()
ott.assert_almost_equal(vol1, base1.getVolume() * extension1.getVolume())
print("Cylinder 1 volume:", vol1)

print("--- Disc-based cylinder 2 ---")
yc2 = 0.8
zc2 = 0.0
R2 = 0.5
H2 = 4.0
f2 = ot.SymbolicFunction(["y", "z"], [f"(y-({yc2}))^2+(z-({zc2}))^2"])
levelSet2 = ot.LevelSet(f2, ot.LessOrEqual(), R2**2)
base2 = ot.LevelSetMesher([N] * 2).build(
    levelSet2, ot.Interval([yc2 - R2, zc2 - R2], [yc2 + R2, zc2 + R2])
)
extension2 = ot.Interval([-H2 / 2] * (dim - 2), [H2 / 2] * (dim - 2))
injection2 = [0] + list(range(3, dim))
cyl2 = otm.Cylinder(base2, extension2, injection2, M)
assert cyl2.isConvex()
vol2 = cyl2.getVolume()
ott.assert_almost_equal(vol2, base2.getVolume() * extension2.getVolume())
print("Cylinder 2 volume:", vol2)

# Test computeMesh on disc cylinders
print("--- computeMesh on disc cylinders ---")
mesh1 = cyl1.computeMesh()
assert mesh1.getVerticesNumber() > 0
assert mesh1.getDimension() == dim
assert mesh1.isValid()
print("Mesh 1 vertices:", mesh1.getVerticesNumber())

mesh2 = cyl2.computeMesh()
assert mesh2.getVerticesNumber() > 0
assert mesh2.getDimension() == dim
assert mesh2.isValid()
print("Mesh 2 vertices:", mesh2.getVerticesNumber())

# Intersect cylinders (existing test)
print("--- Intersect cylinders ---")
inter12 = otm.IntersectionMesher().buildConvex([mesh1, mesh2])
print("Intersection volume:", inter12.getVolume())
ott.assert_almost_equal(inter12.getVolume(), 0.778142671)

print("All tests passed")

File: lib/test/t_Cylinder_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class Cylinder for standard methods
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <cstdio>

#include <openturns/OT.hxx>
#include <openturns/OTtestcode.hxx>

#include "otmeshing/Cylinder.hxx"
#include "otmeshing/IntersectionMesher.hxx"

using namespace OT;
using namespace OT::Test;
using namespace OTMESHING;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    // Test default constructor
    fullprint << "--- Default constructor ---" << std::endl;
    Cylinder cylDefault;
    assert_equal(cylDefault.getDimension(), UnsignedInteger(0));
    fullprint << "Default dimension: " << cylDefault.getDimension() << std::endl;

    // Build a 2D base mesh: square [-1,1]^2 divided into 5x5 cells
    const Indices baseDisc({5, 5});
    const Interval baseInterval(Point({-1.0, -1.0}), Point({1.0, 1.0}));
    const Mesh baseMesh(IntervalMesher(baseDisc).build(baseInterval));
    fullprint << "Base mesh vertices: " << baseMesh.getVerticesNumber() << std::endl;
    fullprint << "Base mesh volume: " << baseMesh.getVolume() << std::endl;

    // Extension: 1D interval [-2, 2]
    const Interval extension(Point({-2.0}), Point({2.0}));
    fullprint << "Extension volume: " << extension.getVolume() << std::endl;

    // Injection: extension maps to dimension 2 (0-indexed)
    const Indices injection({2});

    // Discretization of extension dimension
    const UnsignedInteger M = 3;

    // Test parameter constructor
    fullprint << "--- Parameter constructor ---" << std::endl;
    const Cylinder cylinder(baseMesh, extension, injection, M);
    fullprint << "cylinder: " << cylinder << std::endl;

    // Test getDimension
    fullprint << "Dimension: " << cylinder.getDimension() << std::endl;
    assert_equal(cylinder.getDimension(), UnsignedInteger(3));

    // Test getBase
    fullprint << "--- getBase ---" << std::endl;
    const Mesh retrievedBase = cylinder.getBase();
    assert_almost_equal(retrievedBase.getVertices(), baseMesh.getVertices());
    assert_equal(retrievedBase.getSimplices(), baseMesh.getSimplices());

    // Test getExtension
    fullprint << "--- getExtension ---" << std::endl;
    const Interval retrievedExt = cylinder.getExtension();
    assert_equal(retrievedExt.getLowerBound()[0], extension.getLowerBound()[0]);
    assert_equal(retrievedExt.getUpperBound()[0], extension.getUpperBound()[0]);

    // Test getInjection
    fullprint << "--- getInjection ---" << std::endl;
    const Indices retrievedInj = cylinder.getInjection();
    assert_equal(retrievedInj.getSize(), injection.getSize());
    assert_equal(retrievedInj[0], injection[0]);

    // Test getDiscretization
    fullprint << "--- getDiscretization ---" << std::endl;
    assert_equal(cylinder.getDiscretization(), M);

    // Test getVertices
    fullprint << "--- getVertices ---" << std::endl;
    const Sample vertices = cylinder.getVertices();
    const UnsignedInteger extVerticesNumber = IntervalMesher(Indices({M})).build(extension).getVertices().getSize();
    fullprint << "Base vertices: " << baseMesh.getVerticesNumber()
              << ", ext vertices: " << extVerticesNumber << std::endl;
    assert_equal(vertices.getSize(), baseMesh.getVerticesNumber() * extVerticesNumber);
    assert_equal(vertices.getDimension(), UnsignedInteger(3));

    // Test getBoundingBox
    fullprint << "--- getBoundingBox ---" << std::endl;
    const Interval bbox = cylinder.getBoundingBox();
    fullprint << "BBox: " << bbox << std::endl;
    assert_almost_equal(bbox.getLowerBound()[0], -1.0);
    assert_almost_equal(bbox.getLowerBound()[1], -1.0);
    assert_almost_equal(bbox.getLowerBound()[2], -2.0);
    assert_almost_equal(bbox.getUpperBound()[0],  1.0);
    assert_almost_equal(bbox.getUpperBound()[1],  1.0);
    assert_almost_equal(bbox.getUpperBound()[2],  2.0);

    // Test getVolume: base area (4) * extension length (4) = 16
    fullprint << "--- getVolume ---" << std::endl;
    const Scalar expectedVolume = baseMesh.getVolume() * extension.getVolume();
    fullprint << "Expected volume: " << expectedVolume
              << ", cylinder volume: " << cylinder.getVolume() << std::endl;
    assert_almost_equal(cylinder.getVolume(), expectedVolume);

    // Test isConvex (a square-based cylinder is convex)
    fullprint << "--- isConvex ---" << std::endl;
    assert(cylinder.isConvex());
    fullprint << "Convex: true" << std::endl;

    // Test computeMesh
    fullprint << "--- computeMesh ---" << std::endl;
    const Mesh computedMesh = cylinder.computeMesh();
    fullprint << "Computed mesh vertices: " << computedMesh.getVerticesNumber() << std::endl;
    assert(computedMesh.getVerticesNumber() > 0);

    // Test clone
    fullprint << "--- clone ---" << std::endl;
    Cylinder * clonedCylinder = cylinder.clone();
    assert_equal(clonedCylinder->getDimension(), cylinder.getDimension());
    assert_almost_equal(clonedCylinder->getVolume(), cylinder.getVolume());
    assert_almost_equal(clonedCylinder->getBoundingBox().getLowerBound(),
                        cylinder.getBoundingBox().getLowerBound());
    fullprint << "Clone OK" << std::endl;
    delete clonedCylinder;

    // Test string representation
    fullprint << "--- string representation ---" << std::endl;
    {
      const String repr = cylinder.__repr__();
      fullprint << "__repr__: " << repr << std::endl;
      assert_equal(repr, String("class=Cylinder"));
    }
    {
      const String str = cylinder.__str__();
      fullprint << "__str__: " << str << std::endl;
      assert_equal(str, String("class=Cylinder"));
    }

    // Test save/load via Study
    fullprint << "--- save/load ---" << std::endl;
    const String studyFile("cylinder_test.xml");
    {
      Study study(studyFile);
      study.add("cylinder", cylinder);
      study.save();
    }
    {
      Cylinder loadedCylinder;
      Study loadedStudy(studyFile);
      loadedStudy.load();
      loadedStudy.fillObject("cylinder", loadedCylinder);
      assert_equal(loadedCylinder.getDimension(), cylinder.getDimension());
      assert_almost_equal(loadedCylinder.getVolume(), cylinder.getVolume());
      assert_almost_equal(loadedCylinder.getBoundingBox().getLowerBound(),
                          cylinder.getBoundingBox().getLowerBound());
      assert_almost_equal(loadedCylinder.getBoundingBox().getUpperBound(),
                          cylinder.getBoundingBox().getUpperBound());
      assert_equal(loadedCylinder.getDiscretization(), cylinder.getDiscretization());
    }
    std::remove(studyFile.c_str());
    fullprint << "Save/load OK" << std::endl;

    // Test error cases: invalid injection size (size != extension dimension)
    fullprint << "--- error cases ---" << std::endl;
    {
      const Indices badInjection({2, 3});
      bool caught = false;
      try
      {
        const Cylinder badCyl(baseMesh, extension, badInjection, M);
      }
      catch (const InvalidArgumentException &)
      {
        caught = true;
      }
      assert(caught);
      (void)caught;
      fullprint << "Caught invalid injection size" << std::endl;
    }

    {
      const Indices badInjection2({5});
      bool caught = false;
      try
      {
        const Cylinder badCyl2(baseMesh, extension, badInjection2, M);
      }
      catch (const InvalidArgumentException &)
      {
        caught = true;
      }
      assert(caught);
      (void)caught;
      fullprint << "Caught invalid injection index" << std::endl;
    }

    fullprint << "All tests passed" << std::endl;
  }
  catch (const TestFailed & ex)
  {
    std::cerr << ex.what() << std::endl;
    return ExitCode::Error;
  }
  catch (const std::exception & ex)
  {
    std::cerr << "Unexpected exception: " << ex.what() << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}
